src/sources/journald.rs

```diff
@@ -416,70 +414,91 @@ impl<'a> Batch<'a> {
+        since_now: bool,
+    ) -> Self {
+        Self {
+            path,
+            journal_dir,
+            current_boot_only,
+            since_now,
+        }
+    }
 
-    let stream = FramedRead::new(
-        child.stdout.take().unwrap(),
-        CharacterDelimitedDecoder::new(b'\n'),
-    )
-    .boxed();
+    fn make_command(&self, checkpoint: Option<&str>) -> Command {
+        let mut command = Command::new(&self.path);
+        command.stdout(Stdio::piped());
+        command.arg("--follow");
+        command.arg("--all");
+        command.arg("--show-cursor");
+        command.arg("--output=json");
 
-    let pid = Pid::from_raw(child.id().unwrap() as _);
-    let stop = Box::new(move || {
-        let _ = kill(pid, Signal::SIGTERM);
-    });
+        if let Some(dir) = &self.journal_dir {
+            command.arg(format!("--directory={}", dir.display()));
+        }
 
-    Ok(RunningJournal(stream, stop))
-}
+        if self.current_boot_only {
+            command.arg("--boot");
+        }
 
-fn create_command(
-    path: &Path,
-    journal_dir: Option<&PathBuf>,
-    current_boot_only: bool,
-    since_now: bool,
-    cursor: &Option<String>,
-) -> Command {
-    let mut command = Command::new(path);
-    command.stdout(Stdio::piped());
-    command.arg("--follow");
-    command.arg("--all");
-    command.arg("--show-cursor");
-    command.arg("--output=json");
+        if let Some(cursor) = checkpoint {
+            command.arg(format!("--after-cursor={}", cursor));
+        } else if self.since_now {
+            command.arg("--since=now");
+        } else {
+            // journalctl --follow only outputs a few lines without a starting point
+            command.arg("--since=2000-01-01");
+        }
 
-    if let Some(dir) = journal_dir {
-        command.arg(format!("--directory={}", dir.display()));
+        command
     }
+}
 
-    if current_boot_only {
-        command.arg("--boot");
-    }
+impl StartJournal for StartJournalctl {
+    type Running = RunningJournalctl;
+    fn start(&mut self, checkpoint: Option<&str>) -> crate::Result<(JournalStream, Self::Running)> {
+        let mut command = self.make_command(checkpoint);
 
-    if let Some(cursor) = cursor {
-        command.arg(format!("--after-cursor={}", cursor));
-    } else if since_now {
-        command.arg("--since=now");
-    } else {
-        // journalctl --follow only outputs a few lines without a starting point
-        command.arg("--since=2000-01-01");
+        let mut child = command.spawn().context(JournalctlSpawnSnafu)?;
+
+        let stream = FramedRead::new(
+            child.stdout.take().unwrap(),
+            CharacterDelimitedDecoder::new(b'\n'),
+        )
+        .boxed();
+
+        Ok((stream, RunningJournalctl(child)))
     }
+}
+
+struct RunningJournalctl(Child);
 
-    command
+impl Drop for RunningJournalctl {
+    fn drop(&mut self) {
+        let pid = Pid::from_raw(self.0.id().unwrap() as _);
```

Comment:
It could possibly be marginally safer if we matched on the return of `id()` and only killed if it returned `Some`, since a `None` implies the child is dead anyway. Highly unlikely situation given the context here, so I don't think it's a huge deal if it stays like this..